_src/_posts/2017-01-31-learning-rtb.markdown

```diff
@@ -0,0 +1,204 @@
+vector to a cluster of estimators, each one parameterized according to a grid of
+parameters controlling strength of regularization, number of hashing bits (yes,
+we take advantage of the [hashing trick][trick] as everybody else does),
+estimator memory, etc. Periodically, we evaluate the estimators in the cluster
+using a mix of progressive validation and bootstrap validation in order to
+select the best performing parameterization. The chosen estimator is then
+converted to a tree-like data structure thoroughly optimized for fast model
+evaluation and, finally, it's exported to the bidding subsystem.
+
+We now move on to a more detailed description of the FTRL-P algorithm itself,
+which is the workhorse of our learning system. Let $(x,y)$ denote the input
+vector, with $x$ a set of binary features and $y$ a binary response. Also, let
+$\theta$ denote the vector of model coefficients. Then our prediction at time
+$t$ is computed from the logistic regression model:
+$$
+\hat{y_t} = \frac{1}{1 + e^{-{\theta_t}^T x_t}}
+$$
+Of course, the goal of the algorithm is to learn $\theta_t$ from
+a succession of input events $(x_1,y_1), \ldots, (x_t,y_t)$. For this
+we minimize the sum over each previously seen input event of:
+(i) (The gradient of) the log-likelihood loss, which is a [logloss][] for the
+logistic model;
+(ii) A stabilizing term that introduces some strong convexity into the mix;
+(iii) Regularizing terms that combine lasso and ridge regularization.
+
+After filling in the details and translating everything to Greek:
+$$
+\theta_{t+1} = \underset{\theta}{\operatorname{argmin}}\ 
+{g_{1:t}}^T \theta +
+\frac{1}{2\eta_0}\sum_{s=1}^t ||\theta_s-\theta||^2_{A_s - A_{s-1}} +
+\lambda_1 ||\theta||_1 + \frac{\lambda_2}{2}||\theta||^2_2
+$$
+where:
+* $g_{1:t}$ is the sum of the previous gradients $g_1, \ldots, g_t$ of the
+  loss function.
+* $A_t = (\sum_{s=1}^t g_s g'_s)^{1/2}$, which
+  [minimizes the regret bound][minregret] over Mahalanobis norms for projected
+  gradient descent. That said, in real life we just consider the diagonal matrix
+  $diag(A_t)$ for the sake of computation.
+* $\eta_0$ is the initial learning step.
+* $\lambda_1$ is the strength of the lasso regularization.
+* $\lambda_2$ is the strength of the ridge regularization.
+
+Regularization terms aside, this is mostly equivalent to an online gradient
+descent formulation with learning rate $\eta_t = \eta_0 A_t^{-1}$, but solving
+the minimization in a follow-the-leader fashion allows for an effective
+implementation of lasso regularization that produces coefficients which are
+exactly zero (and, hence, more sparse models that are cheaper to store and
+transfer and, most importantly, way faster to evaluate). Although at first sight
+it might seem that implementing the follow-the-leader update step ---which sums
+over every previous input--- would be harder than implementing the gradient
+descent update step ---which just cares about the last update--- this first
```

Comment:
Missing comma before "this first", to separate both subclauses (still a little loaded for a single sentences, with the  "---" comments separating too much the related "it might seem" and "this first impression").